src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1137 @@
+        min_pin_length_rp_ids: Option<Vec<String>>,
+        pin_auth: Option<Vec<u8>>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if min_pin_length_rp_ids.is_some() {
+            return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+        }
+        if persistent_store.pin_hash().is_some() {
+            match pin_auth {
+                Some(pin_auth) => {
+                    if self.consecutive_pin_mismatches >= 3 {
+                        return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_BLOCKED);
+                    }
+                    let mut message = vec![0xFF; 32];
+                    message.extend(&[0x06, 0x08]);
+                    message.extend(&[min_pin_length as u8, 0x00, 0x00, 0x00]);
+                    // TODO(kaczmarczyck) commented code is useful for the extension
+                    // https://github.com/google/OpenSK/issues/129
+                    // if !cbor::write(cbor_array_vec!(min_pin_length_rp_ids), &mut message) {
+                    //     return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_RESPONSE_CANNOT_WRITE_CBOR);
+                    // }
+                    if !check_pin_auth(&self.pin_uv_auth_token, &message, &pin_auth) {
+                        return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+                    }
+                }
+                None => return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID),
+            };
+        }
+        if min_pin_length < persistent_store.min_pin_length() {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
+        persistent_store.set_min_pin_length(min_pin_length);
+        // TODO(kaczmarczyck) commented code is useful for the extension
+        // https://github.com/google/OpenSK/issues/129
+        // if let Some(min_pin_length_rp_ids) = min_pin_length_rp_ids {
+        //     persistent_store.set_min_pin_length_rp_ids(min_pin_length_rp_ids)?;
+        // }
+        Ok(())
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    fn process_get_pin_uv_auth_token_using_pin_with_permissions(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_hash_enc: Vec<u8>,
+        permissions: u8,
+        permissions_rp_id: Option<String>,
+    ) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        if permissions == 0 {
+            return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+        }
+        // This check is not mentioned protocol steps, but mentioned in a side note.
+        if permissions & 0x03 != 0 && permissions_rp_id.is_none() {
+            return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+        }
+
+        let response =
+            self.process_get_pin_token(rng, persistent_store, key_agreement, pin_hash_enc)?;
+
+        self.permissions = permissions;
+        self.permissions_rp_id = permissions_rp_id;
+
+        Ok(response)
+    }
+
+    pub fn process(
```

Comment:
I think `process_command` or `process_subcommand` would be clearer.

Reply:
I went with `process_subcommand` to distinguish between this function and the one in `mod.rs`.